compiler/src/iree/compiler/Codegen/Common/TileDispatchUsingInterface.cpp

```diff
@@ -266,15 +278,238 @@ static LogicalResult replaceAllStoresWithTiledVersion(
+  SmallVector<int64_t> tileSizes;
+
+  for (auto val : requestedTileSizes) {
+    if (auto constIndexOp = val.getDefiningOp<arith::ConstantIndexOp>()) {
+      auto v = constIndexOp.value();
+      // When tiling is not defined, let's use 1 here.
+      tileSizes.push_back(v == 0 ? 1 : v);
+    } else {
+      return failure();
+    }
+  }
+  return tileSizes;
+}
+
+static SmallVector<AffineMinOp> collectTileSizeMinOps(
+    ArrayRef<OpFoldResult> tileSizes) {
+  SmallVector<AffineMinOp> res;
+
+  for (auto ofr : tileSizes) {
+    if (auto value = ofr.dyn_cast<Value>()) {
+      if (auto minOp = value.getDefiningOp<AffineMinOp>()) {
+        res.push_back(minOp);
+      } else {
+        res.push_back(AffineMinOp());
+      }
+    } else {
+      res.push_back(AffineMinOp());
+    }
+  }
+  return res;
+}
+
+static SmallVector<scf::ForOp> collectNestedLoops(scf::ForOp outerLoop) {
+  SmallVector<scf::ForOp> loops;
+
+  outerLoop.getOperation()->walk<WalkOrder::PreOrder>(
+      [&](scf::ForOp op) { loops.push_back(op); });
+
+  return loops;
+}
+
+// Specialize the tiled distribution loops with the main tile sizes.
+//
+// Inputs:
+//   - TilingResult including main distribution loops and other info
+//
+// Transformed output
+//   cond = (group_id_y != last) && (group_id_x != last)
+//   scf.if cond
+//     distribution loops with static shapes with the tile size
+//   else
+//     distribution loops with dynamic shapes with the tile size
+//
+// Steps:
+//   1. bail out if the loop bounds are already a multiple of the tile size.
+//   2. create a condition for scf.if
+//   3. clone the nested loops in the else block.
+//   4. update the bounded size ops of the original loop nest with the constant
+//      tile size.
+//   5. clone the updated loop nest in the then block.
+static LogicalResult specializeDistributionLoops(TilingResult &tilingResult,
+                                                 PatternRewriter &rewriter) {
+  auto &distLoops = tilingResult.loops;
+  auto tileSizes = getTileSizesAsInt64(tilingResult.tileSizesForLoops);
+  if (failed(tileSizes)) return failure();
+
+  assert(tileSizes->size() == distLoops.size());
+
+  // Check the eligilbility. Unsupported cases are:
+  //   1. Dynamic cases
+  //   2. UB < Tile size
+  //   3. UB == a multiple of the tile size
+  bool isAlreadyMultiple = true;
+  for (auto it : zip(distLoops, *tileSizes)) {
+    auto loop = std::get<0>(it);
+    auto ubOp = loop.getUpperBound().getDefiningOp<arith::ConstantIndexOp>();
+    if (!ubOp) {
+      // TODO: handle dynamic cases by adding more code to check the
+      // conditions.
+      return success();
+    }
+    auto ub = ubOp.value();
+    auto tileSize = std::get<1>(it);
+    if (ub % tileSize != 0) {
+      isAlreadyMultiple = false;
+    }
+    if (ub < tileSize) {
+      return success();
+    }
+  }
+
+  if (isAlreadyMultiple) return success();
+
+  auto minSizeOps = collectTileSizeMinOps(tilingResult.boundedSizesForLoops);
```

Comment:
I'll let Mahesh comment but I believe from discussion an alternative solution would be to emit the specialization directly instead of post processing? Not sure what is better, I'll let Mahesh give more details.

Reply:
Have been going back and forth on this. I think if we are trying to this on `scf.for` itself, with all the fusion etc. to be handled, doing this after the fact might also work, but has more complications.... I am trying to see if there is something we can do that layers better here.